contracts/BC_fusion/StakeHub.sol

```diff
@@ -0,0 +1,1212 @@
+        if (delegateVotePower) {
+            IGovToken(GOV_TOKEN_ADDR).delegateVote(delegator, operatorAddress);
+        }
+    }
+
+    /**
+     * @dev Undelegate BNB from a validator, fund is only claimable few days later
+     * @param operatorAddress the operator address of the validator to be undelegated from
+     * @param shares the shares to be undelegated
+     */
+    function undelegate(
+        address operatorAddress,
+        uint256 shares
+    ) external whenNotPaused notInBlackList validatorExist(operatorAddress) {
+        if (shares == 0) revert ZeroShares();
+
+        address delegator = msg.sender;
+        Validator memory valInfo = _validators[operatorAddress];
+
+        uint256 bnbAmount = IStakeCredit(valInfo.creditContract).undelegate(delegator, shares);
+        emit Undelegated(operatorAddress, delegator, shares, bnbAmount);
+
+        if (delegator == operatorAddress) {
+            _checkValidatorSelfDelegation(operatorAddress);
+        }
+
+        IGovToken(GOV_TOKEN_ADDR).sync(valInfo.creditContract, delegator);
+    }
+
+    /**
+     * @param srcValidator the operator address of the validator to be redelegated from
+     * @param dstValidator the operator address of the validator to be redelegated to
+     * @param shares the shares to be redelegated
+     * @param delegateVotePower whether to delegate vote power to the dstValidator
+     */
+    function redelegate(
+        address srcValidator,
+        address dstValidator,
+        uint256 shares,
+        bool delegateVotePower
+    ) external whenNotPaused notInBlackList validatorExist(srcValidator) validatorExist(dstValidator) {
+        if (shares == 0) revert ZeroShares();
+        if (srcValidator == dstValidator) revert SameValidator();
+
+        address delegator = msg.sender;
+        Validator memory srcValInfo = _validators[srcValidator];
+        Validator memory dstValInfo = _validators[dstValidator];
+        if (dstValInfo.jailed && delegator != dstValidator) revert OnlySelfDelegation();
+
+        _isReceivingFund = 1;
+        uint256 bnbAmount = IStakeCredit(srcValInfo.creditContract).unbond(delegator, shares);
+        if (bnbAmount < minDelegationBNBChange) revert DelegationAmountTooSmall();
+        // check if the srcValidator has enough self delegation
+        if (
+            delegator == srcValidator
+                && IStakeCredit(srcValInfo.creditContract).getPooledBNB(srcValidator) < minSelfDelegationBNB
+        ) {
+            revert SelfDelegationNotEnough();
+        }
+
+        uint256 feeCharge = bnbAmount * redelegateFeeRate / REDELEGATE_FEE_RATE_BASE;
+        (bool success,) = dstValInfo.creditContract.call{ value: feeCharge }("");
+        if (!success) revert TransferFailed();
+
+        bnbAmount -= feeCharge;
+        uint256 newShares = IStakeCredit(dstValInfo.creditContract).delegate{ value: bnbAmount }(delegator);
+        _isReceivingFund = 0;
```

Comment:
We can move this `_isReceivingFund = 0` earlier, right?

Reply:
Done

Reply:
I moved these logic to modifier